### PR TITLE
CDAP-7609 Make Preview store non transactional.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-7609

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  private final Gson gson = new Gson();
+  private final AtomicLong counter = new AtomicLong(0L);
+
+
+  @Inject
+  PreviewTable(LevelDBTableService service) {
+    try {
+      this.service = service;
+      service.ensureTableExists(PREVIEW_TABLE_ID.getDataset());
+      this.table = new LevelDBTableCore(PREVIEW_TABLE_ID.getDataset(), service);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java]: I thought we have different levelDB table per preview run, or not?

Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java]: Yes. We have different data directory per preview run. So this table would get created in the corresponding directory assigned to the preview run.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java]: hum.. but I think the LevelDBTableService is supposed to have one instance per JVM process. How do you configure different directory?

Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java]: @chtyim we made `LevelDBTableService` as non-static and have it singleton per injector. This allows us to create different databases. Each database is opened by single instance of the `LevelDBTableService`. Added second commit to make the `LevelDBTableService.getInstance()` visible for testing only.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java [hunk 2]
```diff
@@ -15,185 +15,44 @@
...
-import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.concurrent.atomic.AtomicLong;
-import javax.annotation.Nullable;
 
 /**
- * Default implementation of the {@link PreviewStore} that stores data in in-memory table
+ * Default implementation of the {@link PreviewStore} that stores data in {@link PreviewTable}.
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/DefaultPreviewStore.java]: what's the point of this class if it just wraps PreviewTable? Is the plan to use PreviewTable somewhere else too?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+  private final Gson gson = new Gson();
+  private final AtomicLong counter = new AtomicLong(0L);
+
+
+  @Inject
+  PreviewTable(LevelDBTableService service) {
+    try {
+      this.service = service;
+      service.ensureTableExists(PREVIEW_TABLE_ID.getDataset());
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java]: Is there any danger of clashing with a user dataset of the same name?

Comment (yaojiefeng) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/preview/PreviewTable.java]: Verified preview run, the level db table is created under each preview run directory, and it is prefixed with namespace name, so there should be no clashing with a user dataset

**Comment by yaojiefeng** [comment]
Build running here: http://builds.cask.co/browse/CDAP-RUT331-1

**Review by albertshau — APPROVED** [review]
lgtm

**Pull request merged**